### DISCUSSION TITLE
Is it possible to get what plugins and core-js features actually used?

### DISCUSSION THREAD

**Discussion by fisker** [babel/babel#13306] [General]
I understand that `@babel/preset-env` support `debug: true`, it will log info on cli, but is it possible to get it as an array? If we can skip actual build will be even better. I want to monitor this info, so it won't include some unwanted transform or polyfills accidentally.

**Comment by nicolo-ribaudo** [comment]
For plugins, there ate two possible ways:

### 1. Using Babel's config loading API

This has the advantage that it loads the preset and resolves all the plugins, so it considers all the options you pass to `@babel/preset-env` (for example, `bugfixes`). However, it might report some internal `preset-env` plugins that are not usable without the preset.

```js
const babel = require("@babel/core");
const presetEnv = require("@babel/preset-env");

// You can avoid specifying the options, and it will automatically load the `babel.config.js` file
const options = babel.loadOptionsSync({
  configFile: false,
  targets: "firefox 60",
  presets: [presetEnv],
});

const usedPlugins = opts.plugins.map(plugin => plugin.key);
```

### 2. Using the `compilation-targets` helper

This has the advantage that you don't rely on `@babel/preset-env`, but only on the compatibility data. However, it's less robust if you want to match `@babel/preset-env`'s behavior since it doesn't support all the options.

```js
const { filterItems } = require("@babel/helper-compilation-targets");
const pluginsData = require("@babel/compat-data");

const plugins = filterItems(
  pluginsData,
  /* @babel/preset-env's include option */ new Set(),
  /* @babel/preset-env's exclude option */ new Set(),
  { firefox: "60.0.0" }
);
```

Or, if you want to specify targets using browserslist' syntax (or a `.browserslistrc` file), you can do
```js
const { default: getTargets, filterItems } = require("@babel/helper-compilation-targets");
const pluginsData = require("@babel/compat-data/plugins");

const plugins = filterItems(
  pluginsData,
  /* include */ new Set(),
  /* exclude */ new Set(),
  getTargets({
    browsers: "firefox 60",
  })
);
```

---

For `core-js`, you can use https://www.npmjs.com/package/core-js-compat.

**Reply by fisker** [reply]
Thanks for your answer.

But, wait, this is not analysing ast? I mean use `useBuiltIns: "usage"`, sorry for not making the request clear.

The easier way I can think, maybe we can accept `debug: function` pass to `@babel/preset-env`? We already print the info during transform. Or pass a `logger`?

**Reply by fisker** [reply]
Or maybe `allow` option?

**Reply by nicolo-ribaudo** [reply]
Oh I see. You could use [`babel-plugin-polyfill-corejs3`](https://github.com/babel/babel-polyfills) directly (which is what `@babel/preset-env` uses internally) and do something like this:
```js
const exclude = new Set([
    "es.array.unscopables.flat-map",
    "es.promise",
    "es.promise.finally",
    "es.string.replace",
    "es.symbol.description",
    "es.typed-array.*",
    "web.*"
]);

config.plugins.push([
  require("babel-plugin-polyfill-corejs3"),
  {
    method: "usage-global", // or "usage-pure", if you want to use `core-js-pure`
    targets,
    shouldInjectPolyfill(name, defaultShouldInject) {
      if (exclude.has(name)) return false;
      if (!defaultShouldInject(name)) return false;

      console.log("This polyfill is used!", name);

      return true;
    }
  }
]);
```

**Reply by nicolo-ribaudo** [reply]
Or maybe
```js
const exclude = new Set([
    "es.array.unscopables.flat-map",
    "es.promise",
    "es.promise.finally",
    "es.string.replace",
    "es.symbol.description",
    "es.typed-array.*",
    "web.*"
]);

const allowed = new Set(["es.array.flat-map", "es.object.from-entries"]);

config.plugins.push([
  require("babel-plugin-polyfill-corejs3"),
  {
    method: "usage-global", // or "usage-pure", if you want to use `core-js-pure`
    targets,
    shouldInjectPolyfill(name, defaultShouldInject) {
      if (exclude.has(name) || !defaultShouldInject(name)) return false;
      if (allowed.has(name)) return true;

      throw new Error("New polyfill detected! " + name);
    }
  }
]);
```

**Reply by fisker** [reply]
Thanks again for the solution, didn't notice there is `shouldInjectPolyfill` option. 

If I understand correctly, `include` will include even it's not needed, `exclude` will exclude even if it's needed, correct?

Do you think my suggestion is reasonable? Add `allow`: it only checked the finnal plugins is in list. Or someone need the opposite `forbid` also only check the list.

If you think we can accept it, I'll wait for that, if not I can try your solution.

**Reply by fisker** [reply]
If it got implemented. In my case, I'll use `exclude` and `allow`, once it fail, I'll add to one of them.

**Reply by nicolo-ribaudo** [reply]
It's correct that `exclude` excludes even if it's needed. On the other hand, `include` will just mark things as unsupported by the target browsers (which could be considered as "include even it's not needed" or not).

For example, if you target Node.js 4 with `method: "usage-global"` (which supports `array.some` but not `Object.fromEntries`) and you have this code:
```js
arr.some(() => true);
```
when you specify `include: "es.array.some"` Babel will include `array.some` (because it's used and, even if it's not needed, the `include` option tells to include it), but it won't include `Object.fromEntries` because, even if it's not supported, it's not used by your code.

> Do you think my suggestion is reasonable? Add allow: it only checked the finnal plugins is in list. Or someone need the opposite fobid also only check the list.

If `shouldInjectPolyfill` works for your usecase I'd prefer to avoid introducing a new option: with `shouldInjectPolyfill` you can already write all the logic you want, so it's more flexible.

**Reply by fisker** [reply]
Okay. thank you!

**Reply by fisker** [reply]
One more question, this only work for polyfills, if I want do the same thing to transform plugins, it won't work, right? Just a question, I don't need it now.

**Reply by nicolo-ribaudo** [reply]
Right, unfortunately for plugins you have to do something like https://github.com/babel/babel/discussions/13306#discussioncomment-733412.

**Reply by fisker** [reply]
Luckly I just removed the plugin from the list yesterday https://github.com/fisker/prettier/blob/d16822e7d77af80f0a086057663067965583b789/scripts/build/bundler.js#L117. Maybe `allow` option is more suitable after all, please reconsider add it in future, no  pressure.

**Reply by nicolo-ribaudo** [reply]
Well, if you were excluding `transform-async-to-generator` then Prettier wasn't working in all the `targets` you specified :sweat_smile:

**Reply by fisker** [reply]
We excluded it since we use `asnyc` the first time one year ago https://github.com/prettier/prettier/pull/6850/files, I removed it only because I found it's not needed anymore...

**Reply by fisker** [reply]
Hello, @nicolo-ribaudo

I just tried your solution, it seems not really the same, I excluded `web.dom-collections.iterator`, but it still get in, maybe because the plugin is not reading result from `shouldInjectPolyfill` or something.

Some polyfills seem not go throw `shouldInjectPolyfill`, maybe because its' required by polyfill self.

Also, can't use `'web.*'`, but this is not big deal.

Code: https://github.com/fisker/prettier/commit/8885b56d27fd2e730b7f7bba1f971d7e2b5dc873


```text
[<CWD>\prettier\node_modules\editorconfig\src\index.js]
The corejs3 polyfill added the following polyfills:
  es.promise { "node":"10" }
  es.symbol.description { "node":"10" }
{ polyfillName: 'es.object.set-prototype-of', shouldInject: false } // passed to shouldInjectPolyfill
es.object.set-prototype-of is excluded // return false
{ polyfillName: 'es.object.create', shouldInject: false }
es.object.create is excluded
```

**Comment by zloirock** [comment]
For polyfills, a tool that could be used for this case is planned for adding in `core-js@4`. Now it's limited by `core-js-compat` and debug mode of `preset-env`.

**Reply by fisker** [reply]
Thank you, hope that can help.

**Comment by fisker** [comment]
This is the case I'm talking about, we are starting use `Array.flatMap()`, `Object.fromEntries()` on Prettier project. But we need support for Node.js 10. After enabling `corejs` option, I found there are unneeded polyfills includes, so I use `exclude` to exclude them now. But I was worried about new syntax/methods will be used, I can't always print debug info and check them, I was hoping get the list and fail the building progress when new polyfills are added. See https://github.com/prettier/prettier/blob/572e0ed90d534234aa40f5c6e6d85df10024cc08/scripts/build/bundler.js#L114-L120
